filecache.go

```diff
@@ -198,47 +352,52 @@ func (c *FileCache) GetFileName(filename string) string {
 	extension := c.DefaultExtension
 
 	// Look in the last 5 characters for a . and extension
-	lastDot := strings.LastIndexByte(filename, '.')
-	if lastDot > len(filename)-6 {
-		extension = filename[lastDot:]
+	lastDot := strings.LastIndexByte(downloadRecord.Path, '.')
+	if lastDot > len(downloadRecord.Path)-6 {
+		extension = downloadRecord.Path[lastDot:]
 	}
 
 	file := fmt.Sprintf("%x%s", hashedFilename, extension)
 	dir := fmt.Sprintf("%x", hashedDir[:1])
 	return filepath.Join(c.BaseDir, dir, filepath.FromSlash(path.Clean("/"+file)))
 }
 
-// onEvictDelete is a callback that is triggered when the LRU cache expires an
-// entry.
-func (c *FileCache) onEvictDelete(key interface{}, value interface{}) {
-	filename := key.(string)
-	storagePath := value.(string)
+// bucketToDownloadManager matches the given bucket to a suitable download manager
+// TODO: Implement this in a more robust / generic way
+func bucketToDownloadManager(bucket string) DownloadManager {
+	switch bucket {
+	case "dropbox":
+		return DownloadMangerDropbox
+	default:
+		return DownloadMangerS3
+	}
+}
 
-	if c.OnEvict != nil {
-		c.OnEvict(key, value)
+// NewDownloadRecord converts the incoming URL path into a download record containing a cached
+// filename (this is the filename on the backing store, not the cached filename locally) and
+// the HTTP request headers that are needed for authentication
+func NewDownloadRecord(r *http.Request) (*DownloadRecord, error) {
```

Comment:
Personally not a fan of passing the `http.Request` in here. That makes this no longer a general purpose library and puts HTTP primitives in. IMO all the HTTP logic should live in Lazyraster and here all you need to worry about are `DownloadRecord`s. You can then let Lazyraster use whatever mechanism it wants to populate the record, including setting the DownloadManager

Reply:
Yeah, that's  good point. Not sure what I was thinking. Probably :sleeping: